### DISCUSSION TITLE
Is it possible to optimize a 3D Delaunay triangulation with CGAL ?

### DISCUSSION THREAD

**Discussion by ImperatorS79** [CGAL/cgal#8497] [Q&A]
I would like to optimize a 3D triangulation with CGAL. More specifically, I would like to be able to, starting for a 3D triangulation object, to optimize the tetrahedrons and respect a size field using Delaunay refinement. Also, only certain tetrahedrons that are "colored" should be optimized, and a certain surface should be preserved.

I know there is a lot of algorithms for this in the different mesh related modules of CGAL. However, they do not seem to be applicable to a simple 3D Delaunay triangulation of points, and I also need to recover the newly added points by CGAL (and to not delete any input points).

**Comment by janetournois** [comment]
Hello,

you can use the optimizers described in the [Tetrahedral mesh generation package](https://doc.cgal.org/latest/Mesh_3/index.html#Mesh_3OptimizationPhase). They move vertices, change the connectivity, and keep surfaces valid, but they do not insert or remove any vertices. Both the input and output triangulations and Regular triangulations (weighted Delaunay triangulation).

You can also use the [Tetrahedral remeshing package](https://doc.cgal.org/latest/Tetrahedral_remeshing/index.html#secTetRemeshing). It does insert and remove vertices, and can be tuned to use a given sizing field, or protect selected vertices to make sure they still are in the output mesh.
The output mesh is not Delaunay anymore, but the surfaces are kept valid too.

What are the criteria you want to fulfill with this optimization?

**Reply by ImperatorS79** [reply]
In th Particle Finite Element Method (PFEM), a fluid is represented with a cloud of points which is Delaunay tetrahedralized (e.g. using the `CGAL::Delaunay_triangulation_3` class). Then various methods are used to "colour" the right fluid tetrahedron and discards the other ones (e.g. it can be an alpha-shape, but it can be other methods too). 

What I would like to do is to optimise this mesh (only the coloured tetrahedra), i.e. the circumcenter-to-shortest-edge, to respect a size field, potentially to remove slivers, using some kind of Delaunay refinement or other optimisations, while keeping the same boundary and being as much Delaunay as possible.

I see there is the `[Remeshing_triangulation_3](https://doc.cgal.org/latest/Tetrahedral_remeshing/classCGAL_1_1Tetrahedral__remeshing_1_1Remeshing__triangulation__3.html)`, which can be used with the Tetrahedral Remeshing package. My questions:
1. Can this triangulation be made as most Delaunay as possible ? 
2. How can I constraint the surface and colour the mesh ?
3. How can I recover the vertex handles added by the tetrahedral remeshing ?

**Reply by janetournois** [reply]
To take benefit of all the new features in tetrahedral remeshing, you should use the master branch, see the documentation [here](https://cgal.geometryfactory.com/CGAL/doc/master/Tetrahedral_remeshing/group__PkgTetrahedralRemeshingRef.html#ga263775c52eeb483a86a16aeb9eb31af0)

1. Not really. There is no such criterion. However the output surface mesh is very isotropic so it must be close to Delaunay, but without any guarantees. You should have a look at the result.
2a. Colors are stored in the `Subdomain_index` of each cell (tetrahedron), in particular in the `Remeshing_cell_base`.
2b. The surfaces (i.e. interfaces between subdomains) are resampled but topologically kept valid. If you want to keep them exactly as is, use `remesh_boundaries=false` and/or `facet_is_constrained_map` to define more surfaces.
3. use `vertex_is_constrained_map` to set vertices that you want to be in the output mesh. The newly inserted vertices will have a `false` value in this property map.

**Reply by ImperatorS79** [reply]
How to send the nodal size to `Adaptive_remeshing_sizing_field` ? I guess I have to provide an info field to the vertex, put I already use it to store the index of my points.

Other question: can I convert a `Delaunay_Triangulation_3` to a `Remeshing_Triangulation_3` ? That way, I would start the optimizations by a Delaunay triangulation.

**Reply by janetournois** [reply]
Sizing is evaluated automatically by `Adaptive_remeshing_sizing_field`, no need to provide more information.

For triangulation conversion, have a look at [convert_to_triangulation_3()](https://cgal.geometryfactory.com/CGAL/doc/master/Tetrahedral_remeshing/group__PkgTetrahedralRemeshingRef.html#gacbc8c23061a2db7654c2ef9209adcfb2) and the provided examples
